torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp

```diff
@@ -416,5 +421,87 @@ GraphExecutorState ProfilingGraphExecutorImpl::getDebugState() {
   return state;
 }
 
+Node* createFallbackGraph(Block* b, ArrayRef<Value*> inputs, Graph* g) {
+  auto graph = std::make_shared<Graph>();
+  auto value_map = [](Value* v) { return v; };
+  graph->block()->cloneFrom(b, value_map);
+
+  auto fallback = g->create(prim::FallbackGraph, inputs, b->outputs().size());
+  fallback->g_(attr::Subgraph, graph);
+
+  for (size_t i = 0; i < b->outputs().size(); i++) {
```

Comment:
Can you please clean up the graph here and remove the old block? you can just iterate from the end of the block -> beginning and destroy every node

Reply:
errr, that doesn't seem like a valid thing to do (not always, at least)
here's how we call this API from `tensorexpr_fuser.cpp`

```cpp
    WithInsertPoint guard(false_block->return_node());
    auto fallback = createFallbackGraph(subgraph->block(), n->inputs(), n->owningGraph());
    n->owningGraph()->insertNode(fallback);
```

where `subgraph` comes from 

```
  void guardFusionGroup(Node* n) {
    auto subgraph = SubgraphUtils::getSubgraph(n);
```

which we shouldn't be mocking with.

Frankly, I think it should be a responsibility of a caller to clean up a graph if required. `createFallbackGraph` (given its name) should just create a fallback graph. We can provide an extra API to clean up a graph.

Reply:
You could just as easily write the same three lines by copying the fallback graph into the block, and then calling `replaceBlockWithFallBackGraph`. I have run into a few places where we leave the graph in a bad state and it's taking me longer to work on the code base as a result. We have 5 duplicated profiles for every node, specialize_autogradzero had the whole graph copied and not cleaned up, I think we should prioritize is this. This is very easy logic to add but makes it easier on pass writers and on debugging the graph. 

Additionally, calling `EliminateDeadCode` is not strictly safe when you're duplicating the block. What if the block has mutation, or a print statement ? These nodes will not get removed by DCE. We need to remove the nodes as part of the API.